src/webgpu/shader/execution/expression/call/builtin/subgroupBallot.spec.ts

```diff
@@ -0,0 +1,202 @@
+  return (1n << BigInt(size)) - 1n;
+}
+
+const kCases = {
+  every_even: {
+    cond: `id % 2 == 0`,
+    filter: (id: number, size: number) => {
+      return id % 2 === 0;
+    },
+    expect: (subgroupSize: number) => {
+      const base = BigInt('0x55555555555555555555555555555555');
+      const mask = getMask(subgroupSize);
+      return base & mask;
+    },
+  },
+  every_odd: {
+    cond: `id % 2 == 1`,
+    filter: (id: number, size: number) => {
+      return id % 2 === 1;
+    },
+    expect: (subgroupSize: number) => {
+      const base = BigInt('0xAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA');
+      const mask = getMask(subgroupSize);
+      return base & mask;
+    },
+  },
+  lower_half: {
+    cond: `id < subgroupSize / 2`,
+    filter: (id: number, size: number) => {
+      return id < Math.floor(size / 2);
+    },
+    expect: (size: number) => {
+      return getMask(Math.floor(size / 2));
+    },
+  },
+  upper_half: {
+    cond: `id >= subgroupSize / 2`,
+    filter: (id: number, size: number) => {
+      return id >= Math.floor(size / 2);
+    },
+    expect: (size: number) => {
+      return getMask(Math.floor(size / 2)) << BigInt(Math.floor(size / 2));
+    },
+  },
+  first_two: {
+    cond: `id == 0 || id == 1`,
+    filter: (id: number) => {
+      return id === 0 || id === 1;
+    },
+    expect: (size: number) => {
+      return getMask(2);
+    },
+  },
+};
+
+function checkBallots(
+  data: Uint32Array,
+  subgroupSize: number,
+  filter: (id: number, s: number) => boolean,
+  expect: (s: number) => bigint
+): Error | undefined {
+  for (let i = 0; i < kInvocations; i++) {
+    const idx = i * 4;
+    let actual = 0n;
+    for (let j = 0; j < 4; j++) {
+      actual |= BigInt(data[idx + j]) << BigInt(32 * j);
+    }
+    let expectedResult = expect(subgroupSize);
+    const subgroupId = i % subgroupSize;
+    if (!filter(subgroupId, subgroupSize)) {
+      expectedResult = 0n;
+    }
+    if (expectedResult !== actual) {
+      return new Error(
+        `Invocation ${i}, subgroup inv id ${
+          i % subgroupSize
+        }\n- expected: ${expectedResult.toString(16)}\n-      got: ${actual.toString(16)}`
+      );
+    }
+  }
+
+  return undefined;
+}
+
+g.test('split')
+  .desc('Tests ballot in a split subgroup')
+  .params(u => u.combine('case', keysOf(kCases)))
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('subgroups' as GPUFeatureName);
+  })
+  .fn(async t => {
+    const testcase = kCases[t.params.case];
+    const wgsl = `
+enable subgroups;
+
+@group(0) @binding(0)
+var<storage, read_write> size : u32;
+
+@group(0) @binding(1)
+var<storage, read_write> output : array<vec4u>;
+
+@compute @workgroup_size(${kInvocations})
+fn main(@builtin(subgroup_size) subgroupSize : u32,
+        @builtin(subgroup_invocation_id) id : u32,
+        @builtin(local_invocation_index) lid : u32) {
+  if (lid == 0) {
+    size = subgroupSize;
+  }
+  if ${testcase.cond} {
+    output[lid] = subgroupBallot();
```

Comment:
This tests the zero-arg ballot.  Add another param for the 1-arg ballot (using true).

Reply:
Adding an unimplemented test.